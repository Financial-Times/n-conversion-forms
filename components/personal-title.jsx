import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export function PersonalTitle ({
	value,
	isDisabled = false,
	hasError = false,
	fieldId = 'title',
	selectId = 'title',
	selectName = 'title',
	options = [
		{ code: 'Mr', description: 'Mr' },
		{ code: 'Mrs', description: 'Mrs' },
		{ code: 'Miss', description: 'Miss' },
		{ code: 'Ms', description: 'Ms' },
	],
	fieldLabel = 'Title',
	isRequired = true,
}) {
	const inputWrapperClassName = classNames([
		'o-forms-input',
		'o-forms-input--select',
		{ 'o-forms-input--invalid': hasError },
	]);

	const fieldTitleClassName = classNames([
		'o-forms-title',
		{ 'o-forms-field--optional': !isRequired },
	]);

	return (
		<label
			id={fieldId}
			className="o-forms-field ncf__validation-error"
			htmlFor={selectId}
		>
			<span className={fieldTitleClassName}>
				<span className="o-forms-title__main">{fieldLabel}</span>
			</span>
			<span className={inputWrapperClassName}>
				<select
					id={selectId}
					name={selectName}
					data-trackable="title"
					aria-required={isRequired}
					required={isRequired}
					disabled={isDisabled}
					defaultValue={value}
				>
					<option value="">Please select a title</option>
					{options.map(({ code, description }) => {
						return (
							<option key={code} value={code}>
								{description}
							</option>
						);
					})}
				</select>
				<span className="o-forms-input__error">
					Please select your personal title
				</span>
			</span>
		</label>
	);
}

PersonalTitle.propTypes = {
	value: PropTypes.string,
	isDisabled: PropTypes.bool,
	hasError: PropTypes.bool,
	fieldId: PropTypes.string,
	selectId: PropTypes.string,
	selectName: PropTypes.string,
	options: PropTypes.arrayOf(
		PropTypes.shape({
			code: PropTypes.string,
			description: PropTypes.string,
		})
	),
	fieldLabel: PropTypes.string,
	isRequired: PropTypes.bool,
};

