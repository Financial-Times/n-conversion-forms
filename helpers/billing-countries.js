import { billingCountries } from 'n-common-static-data';

export { billingCountries };
