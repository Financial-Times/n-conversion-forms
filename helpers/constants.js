const printRegions = {
	cemeaV1: 'CEMEA_V1',
	cemeaV2: 'CEMEA_V2',
	apac: 'APAC',
};

const addressLine1AutoComplete = 'address-line1';
const addressLine2AutoComplete = 'address-line2';
const addressLine3AutoComplete = 'address-line3';

module.exports = {
	printRegions,
	addressLine1AutoComplete,
	addressLine2AutoComplete,
	addressLine3AutoComplete,
};
