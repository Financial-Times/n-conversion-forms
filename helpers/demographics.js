const { demographics } = require('n-common-static-data');

export { demographics };
