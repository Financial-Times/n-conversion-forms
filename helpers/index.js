module.exports = {
	'ncf-countries': require('./ncf-countries'),
	'ncf-common-data': require('./ncf-common-data'),
	cemeaV1ISO: require('./supportedCountries').cemeaV1ISO,
	cemeaV2ISO: require('./supportedCountries').cemeaV2ISO,
	apacISO: require('./supportedCountries').apacISO,
	identifyFTShippingZone: require('./supportedCountries')
		.identifyFTShippingZone,
	countriesSupported: require('./supportedCountries').countriesSupported,
	countriesSupportedISO: require('./supportedCountries').countriesSupportedISO,
	supportedPostcodeValidators: require('./supportedPostcodeValidators')
		.supportedPostcodeValidators,
	allSupportedPostcodeExamples: require('./supportedPostcodeExamples')
		.allSupportedPostcodeExamples,
	demographics: require('./demographics').demographics,
	billingCountries: require('./billing-countries').billingCountries,
};
