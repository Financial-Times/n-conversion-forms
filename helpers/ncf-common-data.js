const _get = require('lodash.get');
const COMMON_DATA = require('n-common-static-data');

module.exports = function ({ hash, fn }) {
	const context = Object.assign({}, { [hash.export]: _get(COMMON_DATA, hash.import) }, this);
	return fn(context);
};
