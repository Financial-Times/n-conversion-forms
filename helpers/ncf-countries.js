const { countries } = require('n-common-static-data').billingCountries;

const MINIMUM_TO_GROUP = 50;
const MINIMUM_TO_SHOW_GROUPS = 2;

module.exports = function ({ hash = {}, fn }) {
	const filter = Array.isArray(hash.filterList) ? hash.filterList : false;
	const value = hash.value;

	let data = countries;

	// Only show countries in the filter
	if (filter) {
		data = data.filter(item => filter.includes(item.code));
	}

	// Select the country
	if (value) {
		data = data.map(item => {
			item.selected = item.code === value;
			return item;
		});
	}

	// Group countries
	if (data.length >= MINIMUM_TO_GROUP) {
		data = groupCountries(data);
	}

	const context = Object.assign({ countries: data }, this);
	return fn(context);
};

/**
 * Produce a frequently used option group
 * @param {Array} countries
 * @return {Array}
 */
function groupCountries (countries) {
	const frequentlyUsed = ['GBR', 'USA', 'JPN', 'FRA', 'CAN'];

	const frequentlyUsedCountries = countries
		.filter(item => frequentlyUsed.includes(item.code))
		.map(item => Object.assign({}, item))
		.sort((a, b) => frequentlyUsed.indexOf(a.code) - frequentlyUsed.indexOf(b.code));

	// Not worth showing return standard countries list
	if (frequentlyUsedCountries.length < MINIMUM_TO_SHOW_GROUPS) {
		return countries;
	}

	// If selected country is in frequently used only leave that one selected
	// otherwise the user is shown the lowest selected option
	const selected = frequentlyUsedCountries.find(item => item.selected);
	if (selected) {
		countries.forEach(item => item.selected = false);
	}

	return [
		{ label: 'Frequently Used', countries: frequentlyUsedCountries },
		{ label: 'Alphabetical', countries }
	];
}
