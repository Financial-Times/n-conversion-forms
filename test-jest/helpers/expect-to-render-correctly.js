import ReactDOMServer from 'react-dom/server';

export const expectToRenderCorrectly = {
	toRenderCorrectly(reactComponent, props) {
		// Render the React component with props.
		const ReactComponentRendered = ReactDOMServer.renderToStaticMarkup(
			reactComponent(props)
		);

		// eslint-disable-next-line jest/no-standalone-expect
		expect(ReactComponentRendered).toMatchSnapshot();

		// This point is reached when the above assertions were successful.
		// The test should therefore always pass, that means it needs to be
		// `true` when used normally, and `false` when `.not` was used.
		return {
			pass: !this.isNot,
		};
	},
};
