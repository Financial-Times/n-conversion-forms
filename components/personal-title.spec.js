import { PersonalTitle } from './index';
import { expectToRenderCorrectly } from '../test-jest/helpers/expect-to-render-correctly';
import { defaultPersonalTitles } from './personal-title';

import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

expect.extend(expectToRenderCorrectly);

describe('PersonalTitle', () => {
	it('render a select with a label', () => {
		const props = {
			options: defaultPersonalTitles
		};

		expect(PersonalTitle).toRenderCorrectly(props);
	});

	it('can render an initial selected value', () => {
		const props = {
			options: defaultPersonalTitles,
			value: 'mr',
		};

		expect(PersonalTitle).toRenderCorrectly(props);
	});

	it('can render a disable select', () => {
		const props = {
			options: defaultPersonalTitles,
			isDisabled: true,
		};

		expect(PersonalTitle).toRenderCorrectly(props);
	});

	it('can render an error message', () => {
		const props = {
			options: defaultPersonalTitles,
			hasError: true,
		};

		expect(PersonalTitle).toRenderCorrectly(props);
	});

	it('can override id and name for select', () => {
		const props = {
			selectId: 'selectId',
			selectName: 'selectName',
		};
		const component = mount(PersonalTitle(props));
		const select = component.find('select#selectId');

		expect(select.exists()).toBe(true);
	});

	it('can override id for field', () => {
		const props = {
			fieldId: 'fieldID',
		};
		const component = mount(PersonalTitle(props));
		const field = component.find('#fieldID');

		expect(field.exists()).toBe(true);
	});

	it('renders with default label wording', () => {
		const props = {};

		const component = mount(PersonalTitle(props));

		expect(component.find('.o-forms-title__main').text()).toEqual(
			'Title'
		);
	});

	it('renders with custom label wording', () => {
		const props = { fieldLabel: 'PersonalTitle' };

		const component = mount(PersonalTitle(props));

		expect(component.find('.o-forms-title__main').text()).toEqual('PersonalTitle');
	});

	it('renders with optional title class, when not required', () => {
		const props = { isRequired: false };
		const component = mount(PersonalTitle(props));

		expect(component.find('.o-forms-title.o-forms-field--optional').length).toEqual(1);
	});
});
