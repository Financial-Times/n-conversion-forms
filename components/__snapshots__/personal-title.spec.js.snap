// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalTitle can render a disable select 1`] = `
<label id="titleField"
       class="o-forms-field ncf__validation-error"
       for="title"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Title
    </span>
  </span>
  <span class="o-forms-input o-forms-input--select">
    <select id="title"
            name="title"
            data-trackable="title"
            aria-required="true"
            required
            disabled
    >
      <option value>
        Please select a title
      </option>
      <option value="Mr">
        Mr
      </option>
      <option value="Mrs">
        Mrs
      </option>
      <option value="Miss">
        Miss
      </option>
      <option value="Ms">
        Ms
      </option>
    </select>
    <span class="o-forms-input__error">
      Please select your personal title
    </span>
  </span>
</label>
`;

exports[`PersonalTitle can render an error message 1`] = `
<label id="titleField"
       class="o-forms-field ncf__validation-error"
       for="title"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Title
    </span>
  </span>
  <span class="o-forms-input o-forms-input--select o-forms-input--invalid">
    <select id="title"
            name="title"
            data-trackable="title"
            aria-required="true"
            required
    >
      <option value>
        Please select a title
      </option>
      <option value="Mr">
        Mr
      </option>
      <option value="Mrs">
        Mrs
      </option>
      <option value="Miss">
        Miss
      </option>
      <option value="Ms">
        Ms
      </option>
    </select>
    <span class="o-forms-input__error">
      Please select your personal title
    </span>
  </span>
</label>
`;

exports[`PersonalTitle can render an initial selected value 1`] = `
<label id="titleField"
       class="o-forms-field ncf__validation-error"
       for="title"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Title
    </span>
  </span>
  <span class="o-forms-input o-forms-input--select">
    <select id="title"
            name="title"
            data-trackable="title"
            aria-required="true"
            required
    >
      <option value>
        Please select a title
      </option>
      <option value="Mr">
        Mr
      </option>
      <option value="Mrs">
        Mrs
      </option>
      <option value="Miss">
        Miss
      </option>
      <option value="Ms">
        Ms
      </option>
    </select>
    <span class="o-forms-input__error">
      Please select your personal title
    </span>
  </span>
</label>
`;

exports[`PersonalTitle render a select with a label 1`] = `
<label id="titleField"
       class="o-forms-field ncf__validation-error"
       for="title"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Title
    </span>
  </span>
  <span class="o-forms-input o-forms-input--select">
    <select id="title"
            name="title"
            data-trackable="title"
            aria-required="true"
            required
    >
      <option value>
        Please select a title
      </option>
      <option value="Mr">
        Mr
      </option>
      <option value="Mrs">
        Mrs
      </option>
      <option value="Miss">
        Miss
      </option>
      <option value="Ms">
        Ms
      </option>
    </select>
    <span class="o-forms-input__error">
      Please select your personal title
    </span>
  </span>
</label>
`;
